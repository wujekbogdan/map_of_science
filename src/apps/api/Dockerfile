FROM node:22.15.0-alpine3.20 AS base

FROM base as pnpm
RUN apk update
RUN apk add --no-cache libc6-compat
ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable

FROM pnpm AS pruner
WORKDIR /app
RUN pnpm add -g turbo
COPY . .
RUN turbo prune @map-of-science/api --docker

FROM pnpm AS builder
WORKDIR /app
COPY --from=pruner /app/out/json/ .
RUN pnpm install --frozen-lockfile
COPY --from=pruner /app/out/full/ .
RUN pnpm turbo build
RUN find . -name node_modules -type d -prune -exec rm -rf '{}' +
RUN pnpm install --prod

FROM base AS runner
WORKDIR /app
RUN addgroup --system --gid 1001 app
RUN adduser --system --uid 1001 app
USER app
COPY --from=builder /app/src/apps/api/dist ./dist
COPY --from=builder /app/src/apps/api/node_modules ./node_modules
COPY --from=builder /app/src/apps/api/package.json ./package.json

CMD node dist/index.js