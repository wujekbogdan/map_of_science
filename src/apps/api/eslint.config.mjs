import { defineNodeConfig } from "@map-of-science/eslint";

export default defineNodeConfig();
