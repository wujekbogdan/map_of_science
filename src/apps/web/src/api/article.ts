const articles: Record<string, () => Promise<{ default: string }>> =
  Object.entries(import.meta.glob("../articles/*.md")).reduce(
    (acc, [path, importFn]) => {
      return {
        ...acc,
        [path.replace("../articles/", "").replace(".md", "")]: importFn,
      };
    },
    {},
  );

const labelTextToLabelId = (label: string) => {
  return label.toLowerCase().replace(/[^a-z0-9]/g, "_"); // Replace non-alphanumeric characters with underscores
};

export const isArticleAvailable = (articleName: string) => {
  const labelId = labelTextToLabelId(articleName);
  return labelId in articles;
};

const markdownToHtml = async (markdown: string) => {
  const { marked } = await import("marked");

  return marked.parse(markdown);
};

export const fetchArticle = async (label: string) => {
  const labelId = labelTextToLabelId(label);

  const path = (await articles[labelId]()).default;
  if (!path) {
    throw new Error(`Article not found: ${labelId}`);
  }

  const response = await fetch(path);
  if (!response.ok) {
    throw new Error(
      `Failed to fetch article: labelId:${labelId}, response: ${response.statusText}`,
    );
  }

  return await markdownToHtml(await response.text());
};
