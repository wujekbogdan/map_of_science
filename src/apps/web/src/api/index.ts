export { loadData } from "./data";
export { fetchArticle } from "./article";
