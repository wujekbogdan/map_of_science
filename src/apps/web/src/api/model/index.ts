import { z as zod } from "zod";

export const ConceptSchema = (z: typeof zod) =>
  z.object({
    index: z.coerce.number(),
    key: z.string(),
  });
export type Concept = zod.infer<ReturnType<typeof ConceptSchema>>;

export const CityLabelSchema = (z: typeof zod) =>
  z
    .object({
      cluster_id: z.coerce.number(),
      label: z.string(),
    })
    .transform((data) => ({
      clusterId: data.cluster_id,
      label: data.label,
    }));
export type CityLabel = zod.infer<ReturnType<typeof CityLabelSchema>>;

export const DataSchema = (z: typeof zod, labels: Map<number, CityLabel>) =>
  z
    .object({
      cluster_id: z.coerce.number(),
      x: z.coerce.number(),
      y: z.coerce.number(),
      num_recent_articles: z.coerce.number(),
      cluster_category: z.coerce.number(),
      growth_rating: z.coerce.number(),
      key_concepts: z.string(),
    })
    .transform((data) => ({
      clusterId: data.cluster_id,
      x: data.x,
      y: -data.y, // To align it with the map coordinate system
      numRecentArticles: data.num_recent_articles,
      clusterCategory: data.cluster_category,
      growthRating: data.growth_rating,
      keyConcepts: data.key_concepts.split(",").map((id) => Number(id)),
      cityLabel: labels.get(data.cluster_id)?.label ?? null,
    }));
export type DataPoint = zod.infer<ReturnType<typeof DataSchema>>;
