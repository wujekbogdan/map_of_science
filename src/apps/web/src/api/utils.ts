import { ZodSchema, ZodTypeDef } from "zod";
import { arrayCollector, mapCollector } from "../csv/collector.ts";
import { createProcessor, withHttpProvider } from "../csv/parse.ts";

type Options<T> = {
  url: string;
  schema: ZodSchema<T, ZodTypeDef, unknown>;
};

export const loadAsArray = async <T>({ url, schema }: Options<T>) => {
  const collector = arrayCollector<T>();
  const processor = createProcessor(schema, collector);

  await withHttpProvider(url, processor.process);

  return processor.getResults();
};

export const loadAsMap = async <T, K>({
  url,
  schema,
  getKey,
}: Options<T> & { getKey: (item: T) => K }) => {
  const collector = mapCollector<K, T>(getKey);
  const processor = createProcessor(schema, collector);

  await withHttpProvider(url, processor.process);

  return processor.getResults();
};
