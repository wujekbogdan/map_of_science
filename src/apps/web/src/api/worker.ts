export const loadData = () => {
  const worker = new ComlinkWorker<typeof import("./data.ts")>(
    new URL("./data.ts", import.meta.url),
  );

  return worker.loadData();
};
