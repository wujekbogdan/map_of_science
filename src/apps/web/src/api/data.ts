import { z } from "zod";
import { DataSchema, ConceptSchema, CityLabelSchema } from "./model";
import { loadAsMap } from "./utils.ts";

// TODO: move this out from here. It does not belong to the API layer.
// It's more of a service layer.
// https://github.com/wujekbogdan/map-of-science/issues/57

/**
 * - City labels loading doesn't depend on anything.
 * - Concepts loading doesn't depend on anything either.
 * - Data points depend only on city labels.
 *
 * To maximize parallelism, we can load city labels and concepts in parallel
 * and then wait for city labels to finish before loading data points.
 */
export const loadData = async () => {
  const loadingConcepts = loadAsMap({
    url: new URL("../../asset/keys.tsv", import.meta.url).href,
    schema: ConceptSchema(z),
    getKey: (item) => item.index,
  });

  const loadingLabels = loadAsMap({
    url: new URL("../../asset/labels.tsv", import.meta.url).href,
    schema: CityLabelSchema(z),
    getKey: (item) => item.clusterId,
  });

  // We're awaiting the labels promise only because at this point we don't care about concepts yet
  const rawLabels = await loadingLabels;

  const [concepts, dataPoints] = await Promise.all([
    loadingConcepts,
    loadAsMap({
      url: new URL("../../asset/data.tsv", import.meta.url).href,
      schema: DataSchema(z, rawLabels),
      getKey: (item) => item.clusterId,
    }),
  ]);

  // Populate labels with x,y coordinates coming from data points
  const labels = [...rawLabels.values()]
    .map(({ clusterId, label }) => {
      const point = dataPoints.get(clusterId);
      return {
        x: point?.x ?? NaN,
        y: point?.y ?? NaN,
        clusterId,
        label,
      };
    })
    .filter(({ x, y }) => !Number.isNaN(x) && !Number.isNaN(y));

  const dataPointsOrdered = new Map(
    [...dataPoints.entries()].sort(
      ([, a], [, b]) => b.numRecentArticles - a.numRecentArticles,
    ),
  );

  return {
    concepts,
    labels,
    dataPoints: dataPointsOrdered,
  };
};
