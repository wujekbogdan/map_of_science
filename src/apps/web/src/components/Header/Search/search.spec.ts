import { describe, it, expect } from "vitest";
import { Concept, DataPoint } from "../../../api/model";
import { createLabelsCollection, search } from "./search.ts";

const map = {
  layer1: {
    attributes: {
      id: "layer1",
      label: "L1",
      style: {
        display: "inline",
      },
    },
    children: [
      {
        path: {
          boundingBox: {
            center: {
              x: -431.8910662922816,
              y: 22.71598460951293,
            },
            max: {
              x: -322.8051709687215,
              y: 258.8283511316347,
            },
            min: {
              x: -540.9769616158417,
              y: -213.39638191260883,
            },
          },
          d: "PLACEHOLDER",
          id: "path186",
          label: "#Symbole-bity-sygnały",
          style: {},
        },
      },
      {
        path: {
          boundingBox: {
            center: {
              x: -175.98698476092548,
              y: 40.46505714629674,
            },
            max: {
              x: -4.789039521850984,
              y: 265.2055539220782,
            },
            min: {
              x: -347.18492999999995,
              y: -184.2754396294847,
            },
          },
          d: "PLACEHOLDER",
          id: "path187",
          label: "#Społeczeństwo",
          style: {},
        },
      },
    ],
  },
  layer2: {
    attributes: {
      id: "layer2",
      label: "L2",
      style: {
        display: "inline",
      },
    },
    children: [
      {
        path: {
          boundingBox: {
            center: {
              x: -390.95909444203926,
              y: 132.81968177008088,
            },
            max: {
              x: -318.7710632026188,
              y: 247.13325685152185,
            },
            min: {
              x: -463.14712568145967,
              y: 18.506106688639907,
            },
          },
          d: "PLACEHOLDER",
          id: "path1",
          label: "#Komputery i komunikacja",
          style: {},
        },
      },
      {
        path: {
          boundingBox: {
            center: {
              x: 220.04397547889005,
              y: 84.34425781473227,
            },
            max: {
              x: 288.4188627868724,
              y: 220.4664682145745,
            },
            min: {
              x: 151.66908817090766,
              y: -51.77795258510995,
            },
          },
          d: "PLACEHOLDER",
          id: "path2",
          label: "#Bakterie i choroby zakaźne",
          style: {},
        },
      },
    ],
  },
  layer3: {
    attributes: {
      id: "layer3",
      label: "L3",
      style: {
        display: "inline",
      },
    },
    groups: [
      {
        attributes: {
          id: "layer4",
          label: "L3-maszyny",
        },
        children: [
          {
            rect: {
              boundingBox: {
                center: {
                  x: -319.7775159,
                  y: -214.0745055,
                },
                max: {
                  x: -316.76377180000003,
                  y: -210.98814099999998,
                },
                min: {
                  x: -322.79126,
                  y: -217.16087,
                },
              },
              height: 6.172729,
              id: "rect31",
              label: "#Urządzenia mechaniczne",
              style: {
                display: "inline",
                fill: "#000000",
              },
              width: 6.0274882,
              x: -322.79126,
              y: -217.16087,
            },
          },
          {
            rect: {
              boundingBox: {
                center: {
                  x: -297.8461759,
                  y: -230.9224255,
                },
                max: {
                  x: -294.8324318,
                  y: -227.836061,
                },
                min: {
                  x: -300.85992,
                  y: -234.00879,
                },
              },
              height: 6.172729,
              id: "rect27",
              label: "#Robotyka",
              style: {
                display: "inline",
                fill: "#000000",
              },
              width: 6.0274882,
              x: -300.85992,
              y: -234.00879,
            },
          },
        ],
      },
      {
        attributes: {
          id: "layer6",
          label: "L3-Ziemia",
        },
        children: [
          {
            rect: {
              boundingBox: {
                center: {
                  x: -141.0597459,
                  y: -293.0270355,
                },
                max: {
                  x: -138.0460018,
                  y: -289.940671,
                },
                min: {
                  x: -144.07349,
                  y: -296.1134,
                },
              },
              height: 6.172729,
              id: "rect32",
              label: "#Asfalt",
              style: {
                display: "inline",
                fill: "#000000",
              },
              width: 6.0274882,
              x: -144.07349,
              y: -296.1134,
            },
          },
        ],
      },
    ],
  },
};

const options = {
  map,
  dataPoints: new Map<number, DataPoint[]>(),
  concepts: new Map<number, Concept[]>(),
} as never; // TODO: Fix this type

// TODO: Add tests for points
// https://github.com/wujekbogdan/map-of-science/issues/60
describe("mapModel", () => {
  it("should map and flatten the model", () => {
    const model = createLabelsCollection(map);
    expect(model).toMatchSnapshot();
  });

  it("should normalize diacritics", () => {
    // GPT-generated phrases in various languages which include diacritics
    const phrases = [
      "zażółć gęślą jaźń", // Polish
      "Élève où tu es allé déjeuner?", // French
      "Fünf köstliche Äpfel überqueren mühelos die Bühne.", // German
      "El niño comió paella en la habitación número cinco.", // Spanish
      "Avó e avô têm um coração cheio de emoção.", // Portuguese
      "İyi bir güneş gökyüzünde yükseliyor.", // Turkish
      "Příliš žluťoučký kůň úpěl ďábelské ódy.", // Czech
      "Őrült görög író újját törte.", // Hungarian
      "Áfram veginn þótt óvinir mæti.", // Icelandic
      "Mâncăm în grădină, lângă tărâmul viselor.", // Romanian
      "Tôi yêu đất nước Việt Nam xinh đẹp.", // Vietnamese
      "Éire agus Gael óg ó Dhún na nGall.", // Irish
    ];

    const el = (phrase: string) => ({
      label: phrase,
      boundingBox: {
        center: {
          x: 0,
          y: 0,
        },
      },
    });

    const model = createLabelsCollection({
      layer1: {
        children: phrases.map((phrase) => ({
          path: el(phrase),
        })),
      },
      layer2: {
        children: phrases.map((phrase) => ({
          path: el(phrase),
        })),
      },
      layer3: {
        groups: [
          {
            children: phrases.map((phrase) => ({
              rect: el(phrase),
            })),
          },
        ],
      },
    } as typeof map); // Cast to avoid providing all properties

    const normalized = [
      "zazolc gesla jazn",
      "eleve ou tu es alle dejeuner?",
      "funf kostliche apfel uberqueren muhelos die buhne.",
      "el nino comio paella en la habitacion numero cinco.",
      "avo e avo tem um coracao cheio de emocao.",
      "i̇yi bir gunes gokyuzunde yukseliyor.",
      "prilis zlutoucky kun upel dabelske ody.",
      "orult gorog iro ujjat torte.",
      "afram veginn thott ovinir maeti.",
      "mancam in gradina, langa taramul viselor.",
      "toi yeu dat nuoc viet nam xinh dep.",
      "eire agus gael og o dhun na ngall.",
    ];
    expect(model.map(({ normalizedLabel }) => normalizedLabel)).toEqual([
      ...normalized,
      ...normalized,
      ...normalized,
    ]);
  });
});

describe("search", () => {
  it("should return search results", () => {
    expect(search(options, "sygnały")).toEqual({
      labels: [
        {
          boundingBox: {
            center: {
              x: -431.8910662922816,
              y: 22.71598460951293,
            },
            max: {
              x: -322.8051709687215,
              y: 258.8283511316347,
            },
            min: {
              x: -540.9769616158417,
              y: -213.39638191260883,
            },
          },
          id: "path186",
          label: "Symbole-bity-sygnały",
          normalizedLabel: "symbole-bity-sygnaly",
        },
      ],
      points: [],
    });

    expect(search(options, "sygnaly")).toEqual({
      labels: [
        {
          boundingBox: {
            center: {
              x: -431.8910662922816,
              y: 22.71598460951293,
            },
            max: {
              x: -322.8051709687215,
              y: 258.8283511316347,
            },
            min: {
              x: -540.9769616158417,
              y: -213.39638191260883,
            },
          },
          id: "path186",
          label: "Symbole-bity-sygnały",
          normalizedLabel: "symbole-bity-sygnaly",
        },
      ],
      points: [],
    });
  });

  it("should return empty array if no results", () => {
    expect(search(options, "nonexistent")).toEqual({
      labels: [],
      points: [],
    });
    expect(search(options, "")).toEqual({
      labels: [],
      points: [],
    });
  });
});
