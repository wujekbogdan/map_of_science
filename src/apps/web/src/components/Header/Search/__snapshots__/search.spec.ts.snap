// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mapModel > should map and flatten the model 1`] = `
[
  {
    "boundingBox": {
      "center": {
        "x": -431.8910662922816,
        "y": 22.71598460951293,
      },
      "max": {
        "x": -322.8051709687215,
        "y": 258.8283511316347,
      },
      "min": {
        "x": -540.9769616158417,
        "y": -213.39638191260883,
      },
    },
    "id": "path186",
    "label": "Symbole-bity-sygnały",
    "normalizedLabel": "symbole-bity-sygnaly",
  },
  {
    "boundingBox": {
      "center": {
        "x": -175.98698476092548,
        "y": 40.46505714629674,
      },
      "max": {
        "x": -4.789039521850984,
        "y": 265.2055539220782,
      },
      "min": {
        "x": -347.18492999999995,
        "y": -184.2754396294847,
      },
    },
    "id": "path187",
    "label": "Społeczeństwo",
    "normalizedLabel": "spoleczenstwo",
  },
  {
    "boundingBox": {
      "center": {
        "x": -390.95909444203926,
        "y": 132.81968177008088,
      },
      "max": {
        "x": -318.7710632026188,
        "y": 247.13325685152185,
      },
      "min": {
        "x": -463.14712568145967,
        "y": 18.506106688639907,
      },
    },
    "id": "path1",
    "label": "Komputery i komunikacja",
    "normalizedLabel": "komputery i komunikacja",
  },
  {
    "boundingBox": {
      "center": {
        "x": 220.04397547889005,
        "y": 84.34425781473227,
      },
      "max": {
        "x": 288.4188627868724,
        "y": 220.4664682145745,
      },
      "min": {
        "x": 151.66908817090766,
        "y": -51.77795258510995,
      },
    },
    "id": "path2",
    "label": "Bakterie i choroby zakaźne",
    "normalizedLabel": "bakterie i choroby zakazne",
  },
  {
    "boundingBox": {
      "center": {
        "x": -319.7775159,
        "y": -214.0745055,
      },
      "max": {
        "x": -316.76377180000003,
        "y": -210.98814099999998,
      },
      "min": {
        "x": -322.79126,
        "y": -217.16087,
      },
    },
    "id": "rect31",
    "label": "Urządzenia mechaniczne",
    "normalizedLabel": "urzadzenia mechaniczne",
  },
  {
    "boundingBox": {
      "center": {
        "x": -297.8461759,
        "y": -230.9224255,
      },
      "max": {
        "x": -294.8324318,
        "y": -227.836061,
      },
      "min": {
        "x": -300.85992,
        "y": -234.00879,
      },
    },
    "id": "rect27",
    "label": "Robotyka",
    "normalizedLabel": "robotyka",
  },
  {
    "boundingBox": {
      "center": {
        "x": -141.0597459,
        "y": -293.0270355,
      },
      "max": {
        "x": -138.0460018,
        "y": -289.940671,
      },
      "min": {
        "x": -144.07349,
        "y": -296.1134,
      },
    },
    "id": "rect32",
    "label": "Asfalt",
    "normalizedLabel": "asfalt",
  },
]
`;
